Make code completions and import suggestions work correctly for extensions with leading using clauses

prolativ opened the pull request:
This is needed after https://github.com/lampepfl/dotty/pull/10940 got merged.

One disputable aspect might be whether an extension method should be suggested in code completion if it requires a given which is not available in scope.
Before the syntax change only trailing using clauses were allowed in extensions, e.g.
```scala
extension (i: Int)(using Foo)
  def xxx = ???
```
and the current behaviour was that for `123.x` `123.xxx` would be suggested even if there's no given `Foo`.
On the one hand not suggesting `xxx` would de-pollute the namespace and cause less confusion if there are multiple extension methods available which are applicable to anything that implements a type class, e.g.
```scala
extension [A](a: A)(using Show[A]) def show: String = ???
```
On the other hand it is still possible to call such methods if you provide the implicit explicitly.
So it seems to keep the current behaviour (to suggest `xxx` instead of hide it) as it is consistent with how code completion currently works with normal methods and methods introduced by implicit classes.

With leading using clauses, e.g.
```scala
extension (using Foo)(i: Int)
  def xxx = ???
```
the situation is slightly different. If a given instance of `Foo` is not in scope it cannot be passed to the method explicitly when one calls it like `123.xxx` so the argument against not suggesting `xxx` doesn't apply anymore.
What's more the type of the receiver parameter might sometimes depend on the leading implicits.
So it makes sense not to suggest `xxx` when there's no given `Foo`. This also seems to be how completion works with implicit classes.

In case of import suggestions this currently seems to work analogically to code completions so I decided to keep this analogy (which also allowed some code reuse) even though sometimes the suggested import might not be enough to make an extension method work

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2133,34 @@ trait Applications extends Compatibility {
-    ref.symbol.is(ExtensionMethod)
-    && !receiver.isBottomType
-    && isApplicableMethodRef(ref, receiver :: Nil, WildcardType)
+  private def tryApplyingReceiverToTruncatedExtMethod(methodSym: TermSymbol, receiver: Tree)(using Context): scala.util.Try[Tree] =
```
Usage of `TermSymbol` cannot handle prefix types, it's better to use `TermRef`.

The name of the method is long, but not informative. I'd suggest make it shorter and add documentation to make its semantics clear.

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2133,34 @@ trait Applications extends Compatibility {
+      case meth: MethodType => meth.newLikeThis(meth.paramNames, meth.paramInfos, defn.AnyType)
+
+    val truncatedSym = methodSym.copy(owner = defn.RootPackage, name = Names.termName(""), info = truncateExtension(methodSym.info))
+    val truncatedRef = ref(truncatedSym).withSpan(Span(0, 0)) // Fake span needed to make this work in REPL
```
Maybe use the position from `receiver`?

prolativ commented on language-server/test/dotty/tools/languageserver/CompletionTest.scala:
```diff
@@ -308,40 +308,140 @@ class CompletionTest {
-          |extension [A](foo: Foo.type) def xxxx: Int = 1
+          |extension (foo: Foo.type) def xxxx[A]: Int = 1
           |object Main { Foo.xx${m1} }""".withSource
       .completion(m1, Set(("xxxx", Method, "[A] => Int")))
```
This is intentional. This is actually a new test case. The old one got copied below and renamed to `completeExtensionMethodFromExtensionWithTypeParameter` because now we need to distinguish between the two

prolativ commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -264,9 +261,12 @@ object Completion {
+        case (symbol, termName) =>
+          if symbol.is(ExtensionMethod) && !qual.tpe.isBottomType then
+            tryApplyingReceiver(symbol).map(_.copy(name = termName))
+          else None
```
Not sure if I got it right. Could you give an example of code for which the current approach wouldn't work? Then I could add it as a test case

liufengyun replied:
Here is one example:

```Scala
trait Foo[T] {
  extension (x: List[T]) foo: List[T] = ???
}

object Bar extends Foo[Int]

import Bar._
```

Now, if we only have the symbol `foo`, we have no way to know that it actually only works for `List[Int]`.

prolativ replied:
Thanks, good catch

prolativ commented:
@liufengyun reworks done.
However your remark about not relying on symbols because they might lose some type information forced me to introduce quite a lot of changes because the problem was already there unspotted before code completion for extension methods was added. On the other hand while working on this I found and fixed a few other issues, especially around how imported definitions were added to completions. I also added quite a lot of tests for that

liufengyun commented:
Nice job, I'll have a look today 👍

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2133,45 @@ trait Applications extends Compatibility {
+    def truncateExtension(tp: Type)(using Context): Type = tp match
+      case poly: PolyType => poly.newLikeThis(poly.paramNames, poly.paramInfos, truncateExtension(poly.resType))
+      case meth: MethodType if meth.isContextualMethod => meth.newLikeThis(meth.paramNames, meth.paramInfos, truncateExtension(meth.resType))
+      case meth: MethodType => meth.newLikeThis(meth.paramNames, meth.paramInfos, defn.AnyType)
```
Nitpick: the lines are too long here (and in some other places), might be good to write the body of case in a new line.

prolativ replied:
done

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2133,45 @@ trait Applications extends Compatibility {
+      case meth: MethodType if meth.isContextualMethod => meth.newLikeThis(meth.paramNames, meth.paramInfos, truncateExtension(meth.resType))
+      case meth: MethodType => meth.newLikeThis(meth.paramNames, meth.paramInfos, defn.AnyType)
+
+    val truncatedSym = methodRef.symbol.asTerm.copy(owner = defn.RootPackage, name = Names.termName(""), info = truncateExtension(methodRef.info))
```
It seems to me `.asTerm` is not useful:

```suggestion
    val truncatedSym = methodRef.symbol.copy(owner = defn.RootPackage, name = Names.termName(""), info = truncateExtension(methodRef.info))
```

prolativ replied:
Actually it is seems to be necessary as for some reason `methodRef.symbol` is just a `Symbol`, not a `TermSymbol`

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2133,45 @@ trait Applications extends Compatibility {
+
+    applyWithoutPostreceiverImplicits(methodRef, receiver)
+      .toOption
+      .map(tree => replaceCallee(tree, ref(methodRef)))
```
I'd suggest merge `tryApplyingReceiver` and `applyWithoutPostreceiverImplicits`, maybe name it `tryApplyExtensionMethod`.

prolativ replied:
I kept them separate for performance as `applyWithoutPostreceiverImplicits` is also reused when searching for potential import suggestions and there we are only interested in whether it was a success, not in the transformed result. Also IMHO with `tryApplyExtensionMethod` it would not be less clear that the arguments following the receiver are not passed to the method

prolativ replied:
If we then move `.toOption` to the body of `tryApplyingExtensionMethod1` then the performance optimization is lost. If we don't care about it (don't we?) then keeping the method split doesn't make sense

liufengyun replied:
I don't think this micro-optimization matters, having a more clear return type is more important. The current result type `Try[Tree]` provides more information than necessary, and can be safely replaced with `Option[Tree]`

liufengyun replied:
BTW, the code can use `try/catch` to return `Option[Tree]` without the overhead of `.toOption`.

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2133,45 @@ trait Applications extends Compatibility {
+
+  def isApplicableExtensionMethod(methodRef: TermRef, receiverType: Type)(using Context): Boolean =
+    methodRef.symbol.is(ExtensionMethod) && !receiverType.isBottomType &&
+      applyWithoutPostreceiverImplicits(methodRef, Typed(EmptyTree, TypeTree(receiverType))).isSuccess
```
What's wrong with `isApplicableMethodRef` in the original code?

prolativ replied:
`isApplicableMethodRef` checks only direct application of arguments without any adjustments like inserting implicits before the supplied arguments list

liufengyun replied:
Thanks for the clarification. Then maybe use `Typyer.extMethodApply` directly on the extension method without any truncating?

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2133,45 @@ trait Applications extends Compatibility {
+   *  A failure is returned if the implicit search fails for any of the leading implicit parameters or if the receiver has a wrong type
+   *  (note that in general the type of the receiver might depend on the exact types of the found instances of the proceding implicits).
+   *  No implicit search is tried for implicits following the receiver or for parameters of the def (D, E).
+   */
```
One question about the specification: if `D` or `E` is not available, maybe we can simply not suggest the extension method to the programmer. This might simplify the code a lot, so that we can simply use `isApplicableMethodRef` as in the original code.

prolativ replied:
As mentioned, `isApplicableMethodRef` doesn't work correctly if there are some leading using sections before the first explicit parameters list. Also as stated in the description of the PR suggesting a method if `D` or `E` are not in scope still makes sense as a user might provide the following implicits explicitly while leading implicits cannot be passed explicitly with dot syntax. This is also consistent with how completions work for implicit classes

liufengyun replied:
Maybe do some formatting of the documentation here.

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -410,23 +412,34 @@ object Completion {
+     * the same `Completion`.
+     */
+    def getCompletions(using Context): List[Completion] = {
+      nameToDenots.toList.groupBy(_._1.toTermName.show).map { (name, namedDenots) =>
```
Can we avoided the `groupBy` here?

prolativ replied:
I don't think so if we want to keep the current semantics although I'm still wondering of this is what we want for the future https://github.com/lampepfl/dotty/issues/11313 but changing this is not an intension of this PR

prolativ commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -135,161 +135,166 @@ object Completion {
+      val imported = grouped.map { (owner, contexts) =>
+        contexts.collect { case context if context.isImportContext =>
+          importedCompletions(using context)
+        }.foldLeft(CompletionScope.empty)(_.mergeDiscardingAmbiguities(_))
```
Actually the previous implementation, which used recursion, couldn't handle some of the use cases well. It's important that imported completions get shadowed by members completions to match how the compiler sees symbols in such situations. I tried to be quite explicit when ambiguities are resolved by shadowing and when using ambiguous symbols would break compilation

liufengyun replied:
> Actually the previous implementation, which used recursion, couldn't handle some of the use cases well.

That is something interesting to know -- conceptually calling the higher-order functions can be rewritten without fusing and inlining. When we have complex chaining with group/ordering, I think writing the code in a plain style will be more maintainable.

prolativ replied:
Not really sure what you mean by "plain style" here but using only one `grouped.map {` block would probably require returning some tricky tuples inside and then unzipping the result to achieve the same semantics

liufengyun replied:
By plain style, I mean try to avoid using high-order functions like `map/fold/filter/group/etc`, and use `while` or recurse methods instead.

liufengyun replied:
I still feel uneasy with the code inside `scopeCompletions`. There are too many chaining of high-order functions. A recursive method or while-loop will be much better in readability and performance.

Could you try to rewrite the code in a plain style to encode the algorithm here? For reference, in the compiler there are many places we use recursion and while-loops to deal with the contex:

https://github.com/lampepfl/dotty/blob/14927459330da0e26f591bd73d330fa6bb198674/compiler/src/dotty/tools/dotc/typer/ImportInfo.scala#L224-L225

prolativ commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +366,28 @@ object Completion {
+          scope
+      }
+
+    extension (scope: CompletionScope.type)
```
That would require making `CompletionScope` an inner class of `CompletionBuffer` so that it has easy access to the name filter defined in it but maybe that's not a problem

prolativ commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -410,23 +412,34 @@ object Completion {
+    val empty = CompletionScope()
+  }
+
+  private case class CompletionScope(nameToDenots: Map[Name, List[SingleDenotation]] = Map.empty) {
```
`CompletionScope` was supposed to be a lightweight wrapper over `Map` taking care of filtering out not matching completions and resolving conflicts between different denotations with the same name.
`CompletionBuffer` holds much more logic

liufengyun replied:
What about keeps a mutable map inside `CompletionBuffer`, which is closer to what the name `buffer` suggests.

liufengyun replied:
To clarify, I'm not proposing the change. I'm just trying to understand the design considerations here.

prolativ replied:
The name comes from the times when it indeed was mutable but some corner cases seemed too hard to fix with mutability. Any idea for a better name?

liufengyun replied:
I'm not familiar with this part of code, so I'm asking questions to learn more about it.

I think the name is OK, I'm more interested in learning the relationships and responsibilities of the classes.

Creating a separate class with only one field looks dubious. However, in this case it may be justified if there is some special semantic operations defined around for this class.

prolativ commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +366,28 @@ object Completion {
+      }
+
+    extension (scope: CompletionScope.type)
+      private def from(denots: Seq[SingleDenotation])(using Context): CompletionScope = {
```
Then it would make sense to also rename `fromNamed` but they can't have the same name because of type erasure

prolativ commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +366,28 @@ object Completion {
     }
+
+    extension (scope: CompletionScope)
+      private def withDenots(denotations: Seq[SingleDenotation], name: Name)(using Context): CompletionScope = {
```
This method is not used anywhere. I just forgot to remove it

prolativ commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -316,53 +323,26 @@ object Completion {
       case site: NamedType if site.symbol.is(Package) =>
         extension (tpe: Type)
-          def accessibleSymbols = tpe.decls.toList.filter(sym => sym.isAccessibleFrom(site, superAccess = false))
+          def accessibleMembers = tpe.allMembers.toList.filter(denot => denot.symbol.isAccessibleFrom(site, superAccess = false))
```
We can't rely on `tpe.decls` because package objects can extend traits with type parameters so, as you had pointed out before, we can't rely on symbols because then we might lose some type information. What I personally still don't understand is why for packages we should extract members of package objects in this package (`packageObjectsDecls`). There was no comment for that and all the tests still pass if I remove the special case for packages

prolativ replied:
It looks like the old implementation was not only unnecessarily complicated but also buggy. I simplified it so that `allMembers` is not necessary

prolativ commented:
@liufengyun any more remarks or are we ready to merge this?

liufengyun commented:
> @liufengyun any more remarks or are we ready to merge this?

I'll have another look soon.

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +365,61 @@ object Completion {
+     *  taking care of filtering out not matching completions
+     *  and enabling easy resolving of conflicts between different denotations with the same name
+     */
+    case class CompletionScope private(nameToDenots: Map[Name, List[SingleDenotation]] = Map.empty) {
```
The documentation does not say whether TermName and TypeName are merged or not.

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +365,61 @@ object Completion {
+          .groupBy(_._1)
+          .collect {
+            case (name, (_, denots) :: Nil)  => name -> denots
+          }
```
It seems to me the code can be simplified with a `fold` or `while`-loop.   The code above creates many intermediate objects and goes through the collection several times.

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +365,61 @@ object Completion {
+      // Merge two scopes using mappings from `that` instead of from `this` in case of name clashes
+      def mergeShadowedBy(that: CompletionScope) = CompletionScope(this.nameToDenots ++ that.nameToDenots)
+
+      // Merge two scopes but discard mappings for names that appear in both scopes
```
```suggestion
      /** Merge two scopes but discard mappings for names that appear in both scopes */
```

Will/should a TermName `a` considered the same as a TypeName `a`?

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +365,61 @@ object Completion {
+     */
+    case class CompletionScope private(nameToDenots: Map[Name, List[SingleDenotation]] = Map.empty) {
+
+      // Merge two scopes using mappings from `that` instead of from `this` in case of name clashes
```
```suggestion
      /** Merge two scopes using mappings from `that` instead of from `this` in case of name clashes */
```

Will/should a TermName `a` considered the same as a TypeName `a`?

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +365,61 @@ object Completion {
+          .filter((name, denot) => include(denot.symbol, name))
+          .toList
+          .groupBy(_._1)
+          .map((name, namedDenots) => name.stripModuleClassSuffix -> namedDenots.map(_._2))
```
Could you please explain a little bit about `namedDenots.map(_._2)`? The code might be simpler with a simple loop or `fold` on `namedDenotations `.

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +366,28 @@ object Completion {
+          CompletionScope(scope.nameToDenots + (shortName -> denots.toList))
+        else
+          scope
+      }
```
It seems to me the code can be simplified with a fold or while-loop.

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +365,61 @@ object Completion {
+        CompletionScope(mappings)
+      }
+
+      def fromNamed(namedDenotations: Seq[(Name, SingleDenotation)])(using Context): CompletionScope = {
```
From the type `Seq[(Name, SingleDenotation)]`, does it mean it's possible that from some tuple `(Name, SingleDenotation)`, the name and the `denot.name` is different?

prolativ replied:
Yes, for renaming imports like `import Foo.{xxx => yyy}`

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +365,61 @@ object Completion {
+          val desc = description(typesFirst)
+          Completion(name, desc, typesFirst.map(_.symbol))
+        }.toList
+      }
```
It seems we only care about symbols here, can we simply change `SingleDenotation` to `Symbol` in the map?

prolativ replied:
We need denotations to get the exact type information

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +365,61 @@ object Completion {
+    }
+
+    /** A wrapper over a map from names to their denotations
+     *  taking care of filtering out not matching completions
```
I cannot find corresponding logic about `filtering` in the class. It is not the responsibility of the class?

liufengyun commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -386,6 +365,61 @@ object Completion {
+       * the same `Completion`.
+       */
+      def getCompletions(using Context): List[Completion] = {
+        nameToDenots.toList.groupBy(_._1.toTermName.show).map { (name, namedDenots) =>
```
```suggestion
        nameToDenots.toList.groupBy(_._1.toTermName).map { (name, namedDenots) =>
```

Name comparison is much faster than String.

Just curious, can we simply normalize the keys to TermName beforehand? This is done previously in `RenameAwareScope`.

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2135,49 @@ trait Applications extends Compatibility {
+   *  No implicit search is tried for implicits following the receiver or for parameters of the def (D, E).
+   */
+  def tryApplyingExtensionMethod(methodRef: TermRef, receiver: Tree)(using Context): Option[Tree] =
+      // Drop all parameters sections of an extension method following the receiver; the return type after truncation is not important
```
```suggestion
    // Drop all parameters sections of an extension method following the receiver
    // The return type after truncation is not important
```

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2135,49 @@ trait Applications extends Compatibility {
+      case TypeApply(fun, args) => TypeApply(replaceCallee(fun, replacement), args)
+      case _: Ident => replacement
+
+    val truncatedSym = methodRef.symbol.asTerm.copy(owner = defn.RootPackage, name = Names.termName(""), info = truncateExtension(methodRef.info))
```
```suggestion
    val truncatedSym = methodRef.symbol.asTerm.copy(info = truncateExtension(methodRef.info))
```

It seems the name and owner don't matter here.

liufengyun commented on compiler/src/dotty/tools/dotc/typer/Applications.scala:
```diff
@@ -2133,8 +2135,49 @@ trait Applications extends Compatibility {
+
+  def isApplicableExtensionMethod(methodRef: TermRef, receiverType: Type)(using Context): Boolean =
+    methodRef.symbol.is(ExtensionMethod) && !receiverType.isBottomType &&
+      tryApplyingExtensionMethod(methodRef, Typed(EmptyTree, TypeTree(receiverType))).nonEmpty
```
It's better to avoid `EmptyTree` inside expressions:

```suggestion
      tryApplyingExtensionMethod(methodRef, nullLiteral.asInstance(receiverType))).nonEmpty
```

prolativ commented:
@liufengyun it took me quite some time but I think I managed to significantly simplify the code in many places while also handling some additional corner cases.
`ScopeCompletions` class turned out to be both too heavy and not flexible enough for some cases so I switched to using ordinary `Map`s and performing the filtering exactly where it's needed.
If the implementation of `scopeCompletions` still seems confusing to you it might be simply because the logic there is complex by itself so I tried to at least document it well.

liufengyun commented:
@prolativ Thanks a lot for taking the effort to improve the PR. I'll review and get it merged soon.

liufengyun reviewed (approved):
LGTM, well done 🎉

prolativ commented on compiler/src/dotty/tools/dotc/interactive/Completion.scala:
```diff
@@ -366,62 +423,23 @@ object Completion {
+   */
+  private case class ScopedDenotations(
+    denots: Seq[SingleDenotation],
+    scope: Scope,
```
The point is that multiple contexts can share the same scope e.g. when you do multiple imports one after another
```
import Foo.{xxx => zzz}
import Bar.{yyy => zzz}
```
and scope seems to be the only part  of a context that seems to matter here for the distinction

liufengyun replied:
I was not clear in my previous message. I meant to use `ctx` here and use `ctx.scope` in the comparison.
